vcd/resource_vcd_nsxt_alb_virtual_service_req_rules.go

```diff
@@ -0,0 +1,783 @@
+				ForceNew:    true,
+				Description: "NSX-T ALB Virtual Service ID",
+			},
+			"rule": {
+				Type:        schema.TypeList,
+				Required:    true,
+				Elem:        nsxtAlbVirtualServiceReqRule,
+				Description: "A single HTTP Request Rule",
+			},
+		},
+	}
+}
+
+var nsxtAlbVirtualServiceReqRule = &schema.Resource{
+	Schema: map[string]*schema.Schema{
+		"name": {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: "Name of the rule",
+		},
+		"active": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     true,
+			Description: "Defines if the rule is active or not",
+		},
+		"logging": {
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Description: "Defines whether to enable logging with headers on rule match or not",
+		},
+		"match_criteria": {
+			Type:        schema.TypeSet,
+			MaxItems:    1,
+			Required:    true,
+			Description: "Rule matching Criteria",
+			Elem:        nsxtAlbVsReqAndSecRuleMatchCriteria,
+		},
+		"actions": {
+			Type:        schema.TypeSet,
+			MaxItems:    1,
+			Required:    true,
+			Description: "Actions to perform with the rule that matches",
+			Elem:        nsxtAlbVirtualServiceReqRuleActions,
+		},
+	},
+}
+
+var nsxtAlbVsReqAndSecRuleMatchCriteria = &schema.Resource{
+	Schema: map[string]*schema.Schema{
+		"client_ip_address": {
+			Type:        schema.TypeList,
+			MaxItems:    1,
+			Optional:    true,
+			Description: "Client IP Address criteria",
+			Elem: &schema.Resource{
+				Schema: map[string]*schema.Schema{
+					"criteria": {
+						Type:         schema.TypeString,
+						Required:     true,
+						ValidateFunc: validation.StringInSlice([]string{"IS_IN", "IS_NOT_IN"}, false),
+						Description:  "Criteria to use for IP address matching the HTTP request. Options - IS_IN, IS_NOT_IN.",
+					},
+					"ip_addresses": {
+						Type:        schema.TypeSet,
+						Required:    true,
+						Description: "A set of IP addresses",
+						Elem: &schema.Schema{
+							Type: schema.TypeString,
+						},
+					},
+				},
+			},
+		},
+		"service_ports": {
+			Type:        schema.TypeList,
+			MaxItems:    1,
+			Optional:    true,
+			Description: "Service Port criteria",
+			Elem: &schema.Resource{
+				Schema: map[string]*schema.Schema{
+					"criteria": {
+						Type:         schema.TypeString,
+						Required:     true,
+						ValidateFunc: validation.StringInSlice([]string{"IS_IN", "IS_NOT_IN"}, false),
+						Description:  "Criteria to use for IP address matching the HTTP request. Options - IS_IN, IS_NOT_IN",
+					},
+					"ports": {
+						Type:        schema.TypeSet,
+						Required:    true,
+						Description: "A set of TCP ports. Allowed values are 1-65535",
+						Elem: &schema.Schema{
+							Type:             schema.TypeInt,
+							ValidateDiagFunc: validation.ToDiagFunc(validation.IntBetween(1, 65535)),
+						},
+					},
+				},
+			},
+		},
+		"protocol_type": {
+			Type:         schema.TypeString,
+			Optional:     true,
```

Comment:
Just a doubt, should this have a Default? Or be Required.

Reply:
No, this one is "one of the options" It is adjacent to the other more nested blocks and is up to user to use it or not in the rule.